### PR TITLE
Review of changes that are required for InfluxDB Reactive Java client

### PR DESCRIPTION
**Pull request by bednar**
This PR solve [#392](https://github.com/influxdata/influxdb-java/issues/392) by adding possibility to reuse influxdb-java client as a core part of [influxdb-java-reactive](https://github.com/bonitoo-io/influxdb-java-reactive) - **Reactive Java client for InfluxDB**.

#### Summary of changes
1. The `InfluxDBImpl` constructor has parameter for customise `Retrofit.Builder`. That builder is use in **influxdb-java-reactive** client for create reactive version of `InfluxDBService`.
2. The `InfluxDBResultMapper` is able to handle results which a different time precision. This issue is described below
3. The `InfluxDBResultMapper` is able to map `Integer` to `Instant`

#### Precision issue

The `InfluxDBResultMapper` is not able map QueryResults with different precision than Milliseconds.

```java
@Test
void testToPOJO_Precision() {

	String database = "precision_database";
	influxDB.deleteDatabase(database);
	influxDB.createDatabase(database);

	// write in Nanoseconds
	influxDB.write(database, "autogen", InfluxDB.ConsistencyLevel.ONE,
				"h2o_feet,location=coyote_creek level\\ description=\"below 3 feet\",water_level=2.927 1000000000");

	// query in Seconds
	QueryResult queryResults = influxDB.query(new Query("select * from h2o_feet", database), TimeUnit.SECONDS);

	// Mapper expected time in Milliseconds
	List<H2OFeetMeasurement> measurements = new InfluxDBResultMapper().toPOJO(queryResults, H2OFeetMeasurement.class);

	Assertions.assertEquals(1, measurements.size());
	Assertions.assertEquals("coyote_creek", measurements.get(0).location);
	Assertions.assertEquals(2.927D, (double) measurements.get(0).level);
	Assertions.assertEquals("below 3 feet", measurements.get(0).description);
	Assertions.assertEquals(Instant.ofEpochSecond(1), measurements.get(0).time);
}

@Measurement(name = "h2o_feet", timeUnit = TimeUnit.NANOSECONDS)
public static class H2OFeetMeasurement {

	@Column(name = "location", tag = true)
	private String location;

	@Column(name = "water_level")
	private Double level;

	@Column(name = "level description")
	private String description;

	@Column(name = "time")
	private Instant time;
}
```

---
[codecov](https://codecov.io/gh/bonitoo-io/influxdb-java/branch/reactive-pull-request)

@lxhoan, @ivankudibal , @dubsky, @rhajek can you look at this?

### CODE REVIEW

**Comment by lxhoan** [comment]
There are some points to fix but for now I only want to put one comment from architecture POV, that we should try to use Delegation rather than Inheritance, so you don't need to make such a big change to influxdb-java (If I were influxdb-java maintainer, I would ask what benefit this PR can return in comparison to the risk it can cause)

The only thing I think we need to change in influxdb-java to support for a Reactive client (your implementation) that an InfluxDBImpl constructor to receive a custom Retrofit builder

**Comment by bednar** [comment]
Thx @lxhoan, It's a good advice and I will look at it.

**Comment by bednar** [comment]
@lxhoan you had right :+1:

 Now the PR is more simple. Thanks for advice and please check the sources.

**Comment by lxhoan** [comment]
hi @bednar 

It now seems OK for me. One more thing, even though it's trivial in my opinion, but maybe distract the reviewers, that I see you tried to refactor some code, mostly replace inline constants by class constant properties, if those refactors are not necessary I really suggest you to keep it as it it, so the reviewers can stay more focused on the scope of this PR

(I think you want to improve code quality however we can consider it in separate PR)

**Comment by bednar** [comment]
@lxhoan Thank you very much for your advices...

You have more experience with maintainers of influxdb-java so I will follow your suggestions and I will remove refactoring to constants. 

Thanks you again

**Comment by lxhoan** [comment]
NP,
Without those refactors I see you have reduced more than a half of changed files

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 1]
```diff
@@ -146,7 +169,7 @@ public InfluxDBImpl(final String url, final String username, final String passwo
       break;
     }
 
-    this.retrofit = new Retrofit.Builder().baseUrl(url).client(
+    this.retrofit = retrofitBuilder.baseUrl(url).client(
```
Comment (fmachado) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: Just a heads up: you may have here `ConcurrentModificationException` as you are adding a `converterFactory` to a shared and non thread-safe `retrofitBuilder`. Worst case scenatio if you don't trigger a CME, you will end up having multiple `converterFactory` [added to the internal list in retrofitBuilder](https://github.com/square/retrofit/blob/master/retrofit/src/main/java/retrofit2/Retrofit.java#L515).

File: src/main/java/org/influxdb/impl/InfluxDBResultMapper.java [hunk 2]
```diff
@@ -152,7 +202,7 @@ void throwExceptionIfResultWithError(final QueryResult queryResult) {
     });
   }
 
-  void cacheMeasurementClass(final Class<?>... classVarAgrs) {
+  public void cacheMeasurementClass(final Class<?>... classVarAgrs) {
```
Comment (fmachado) [inline src/main/java/org/influxdb/impl/InfluxDBResultMapper.java]: Please revert the access modifier from this method on this class and from all methods from other classes to the minimum required. There is no need to make this method public.

File: src/main/java/org/influxdb/impl/InfluxDBResultMapper.java [hunk 3]
```diff
@@ -172,13 +222,22 @@ void cacheMeasurementClass(final Class<?>... classVarAgrs) {
     }
   }
 
-  String getMeasurementName(final Class<?> clazz) {
+  public String getMeasurementName(final Class<?> clazz) {
     return ((Measurement) clazz.getAnnotation(Measurement.class)).name();
   }
 
+  public <T> ConcurrentMap<String, Field> getColNameAndFieldMap(final Class<T> clazz) {
```
Comment (fmachado) [inline src/main/java/org/influxdb/impl/InfluxDBResultMapper.java]: There is no need to have this method: you are not mocking it and now you are exposing an internal structure publicly to the external code, allowing it to be changed. **Please, revert this change.**

**Comment by fmachado** [comment]
Oh, great! Now I'm reviewing code that is not from influxdb-java! :-D My apologies guys! :-D Please, do not consider my comments.

**Comment by fmachado** [comment]
I'll add the proper comments on the correct PR.
